Maritime tutorials 💧 - Part 3 of 4 (#2259)
gazebosim/gz-sim: tutorials/lander.md

\page lander

# Overview

This tutorial explains how to create and load an underwater lander vehicle in
Gazebo. This type of vehicles don't usually have control surfaces or thrusters.
They typically sink until they reach the seafloor or certain depth and go back
to the surface while they collect sensor measurements.

## Related tutorials

https://gazebosim.org/api/sim/8/create_vehicle.html

https://gazebosim.org/api/sim/8/adding_visuals.html

https://gazebosim.org/api/sim/8/frame_reference.html

https://gazebosim.org/api/sim/8/adding_system_plugins.html

https://gazebosim.org/api/sim/8/theory_buoyancy.html

https://gazebosim.org/api/sim/8/theory_hydrodynamics.html

# Create your vehicle

As an example, we'll go through the process of creating an Inkfish underwater
lander vehicle.

## Basic SDF

Create a workspace to store your brand new model named `my_lander`.

```bash
mkdir -p ~/gazebo_maritime/models/my_lander && cd ~/gazebo_maritime/models/my_lander
```

Download the `model.config` file and copy it within that directory:

```bash

> wget https://raw.githubusercontent.com/gazebosim/gz-sim/main/tutorials/files/lander/model.config -o ~/gazebo_maritime/models/my_lander

Reviewer: ```suggestion
wget https://raw.githubusercontent.com/gazebosim/gz-sim/main/tutorials/files/lander/model.config -o ~/gazebo_maritime/models/my_lander/model.config
```

```

In its simple version, the lander does not have any moving pieces, so the SDF
model only has one single link, `base_link` in our case.

Create a `model.sdf` and paste the following content:

```xml
<?xml version="1.0"?>
<sdf version="1.6">
  <model name="lander">
    <link name="base_link">
      <inertial>
        <!-- Center of mass -->
        <pose>0 0 -0.982494 0 0 0</pose>

        <mass>1209.175</mass>
        <inertia>
          <ixx>205.227945746</ixx>
          <ixy>0</ixy>
          <ixz>0</ixz>
          <iyy>412.254844006</iyy>
          <iyz>0</iyz>
          <izz>383.659583213</izz>
        </inertia>
      </inertial>

      <visual name="chassis_visual">
        <geometry>
          <box>
            <size>1.325 1.1503 1.5</size>
          </box>
        </geometry>
      </visual>
    </link>

  </model>

> </sdf>
> 

Reviewer 2: ```suggestion
</sdf>
```

Author: Line removed in [2648ff4](https://github.com/gazebosim/gz-sim/pull/2259/commits/2648ff4f45e479dc7636bb6ff6cce1f4db6e2b0e).

```

We can now test it in Gazebo. Launch it with:

```bash
gz sim ~/gazebo_maritime/models/my_lander/model.sdf
```

> You should visualize your model as a box.

Reviewer 3: ```suggestion
You should see your model visualized as a box.
```

@image html files/lander/simple_lander.png

## Adding visuals

Let's now add nicer-looking visuals to our lander. Download the following
`COLLADA` mesh file from here.

```bash
mkdir ~/gazebo_maritime/models/my_lander/meshes

> wget https://raw.githubusercontent.com/gazebosim/gz-sim/main/tutorials/files/lander/inkfish-lander.dae -o ~/gazebo_maritime/models/my_lander/meshes

Reviewer: ```suggestion
wget https://raw.githubusercontent.com/gazebosim/gz-sim/main/tutorials/files/lander/inkfish-lander.dae -o ~/gazebo_maritime/models/my_lander/meshes/inkfish-lander.dae
```

```

Replace the `chassis_visual` element in your `model.sdf` with the following
block:

```xml
<visual name="chassis_visual">
  <geometry>
    <mesh>
      <uri>meshes/inkfish-lander.dae</uri>
    </mesh>
  </geometry>
</visual>
```

And launch Gazebo to see the results:

```bash
gz sim ~/my_models/my_lander/model.sdf
```

Your box should now be replaced with a better looking mesh.

@image html files/lander/lander_visuals.png

## Checking frame of reference

Launch your lander in Gazebo as we did before:

```bash
gz sim ~/gazebo_maritime/models/my_lander/model.sdf
```

You can visualize the frame of reference of your vehicle by loading the plugin
`Transform Control` from the GUI.

@image html files/lander/gui_transform_control.png

Then, click on the vehicle, and then, click on the `Translate mode` icon. You'll
now visualize your lander frame of reference. As this is the expected frame of
reference, we won't need to apply changes in the COLLADA mesh or add any
`<pose>` offset to our `model.sdf`.

## Adding plugins

### Buoyancy

We are going to start extending our model adding buoyancy properties. For that,

> we need to do two things:
> 
> 
> 1. Create a Gazebo world SDF file and load the buoyancy world plugin.

Reviewer 2: ```suggestion
we need to do two things:

1. Create a Gazebo world SDF file and load the buoyancy world plugin.
```

```

Hit play and the lander will slowly sink and hit the seafloor.

### Hydrodynamics

As an underwater vehicle, the lander should be influenced by the water as it
moves. We can do that by attaching the `hydrodynamics` plugin to our lander.
Add the next SDF block to your `model.sdf`. Note that we have empirically
adjusted its values. Follow the
[hydrodynamics tutorial](https://gazebosim.org/api/sim/8/theory_hydrodynamics.html)
for recommendations about how to tune its values.

```xml
<!-- Hydrodynamics -->    
<plugin
  filename="gz-sim-hydrodynamics-system"
  name="gz::sim::systems::Hydrodynamics">
  <link_name>base_link</link_name>
  <xDotU>-4.876161</xDotU>
  <yDotV>-126.324739</yDotV>
  <zDotW>-126.324739</zDotW>
  <kDotP>0</kDotP>
  <mDotQ>-33.46</mDotQ>
  <nDotR>-33.46</nDotR>
  <xUabsU>-6.2282</xUabsU>
  <xU>0</xU>
  <yVabsV>-601.27</yVabsV>
  <yV>0</yV>
  <zWabsW>-601.27</zWabsW>
  <zW>0</zW>
  <kPabsP>-0.1916</kPabsP>
  <kP>0</kP>
  <mQabsQ>-632.698957</mQabsQ>
  <mQ>0</mQ>
  <nRabsR>-632.698957</nRabsR>
  <nR>0</nR>
</plugin>
```

### Drop weight

A common feature in maritime landers is to include a drop weight device. When

> this device is enabled, some part of the vehicle is dettached from the main

Reviewer 3: ```suggestion
this device is enabled, some part of the vehicle is detached from the main
```

chassis. At this point, the vehicle becomes positive buoyant, and starts moving
up towards the water surface.

> In order to simulate this behavior, we'll need to make two changes to our lander

Reviewer 3: some places use `behaviour` while some `behavior`. Keep them consistent?

Author: Updated in [f667c7e](https://github.com/gazebosim/gz-sim/pull/2259/commits/f667c7e279672cee3b39b5b17fc4162ed20a5929).
